src/lair/toplevel.rs

```diff
@@ -189,18 +267,30 @@ impl<F: Field + Ord> FuncE<F> {
 }
 
 impl<F: Field + Ord> BlockE<F> {
-    fn check_and_link_with_ops<C1: Chipset<F>, C2: Chipset<F>>(
-        &self,
-        mut ops: Vec<Op<F>>,
-        ctx: &mut CheckCtx<'_, C1, C2>,
-    ) -> Block<F> {
-        self.ops
-            .iter()
-            .for_each(|op| op.check_and_link(&mut ops, ctx));
+    fn check<C1: Chipset<F>, C2: Chipset<F>>(&self, ctx: &mut CheckCtx<'_, C1, C2>) {
+        self.ops.iter().for_each(|op| op.check(ctx));
+        self.ctrl.check(ctx);
+    }
+
+    fn expand(&self, ctx: &mut ExpandCtx) -> BlockE<F> {
+        self.expand_with_ops(vec![], ctx)
+    }
+
+    fn expand_with_ops(&self, mut ops: Vec<OpE<F>>, ctx: &mut ExpandCtx) -> BlockE<F> {
+        self.ops.iter().for_each(|op| op.expand(&mut ops, ctx));
+        let ops = ops.into();
+        let ctrl = self.ctrl.expand(ctx);
+        BlockE { ops, ctrl }
+    }
+
+    fn compile<C1: Chipset<F>, C2: Chipset<F>>(&self, ctx: &mut LinkCtx<'_, C1, C2>) -> Block<F> {
+        let mut ops = vec![];
+        self.ops.iter().for_each(|op| op.compile(&mut ops, ctx));
         let ops = ops.into();
         let saved_return_idents = std::mem::take(&mut ctx.return_idents);
-        let ctrl = self.ctrl.check_and_link(ctx);
+        let ctrl = self.ctrl.compile(ctx);
         let block_return_idents = std::mem::take(&mut ctx.return_idents);
+        // sanity check
```

Comment:
nit: Can this sanity check (the assert below this comment) be moved to the `check` function on `BlockE` instead?